File: Server_VPS/api/src/routes/telemetry.js
const express = require('express');
const db = require('../services/database');
const { authMiddleware, checkDeviceOwnership } = require('../middleware/auth');
const router = express.Router();

// Get telemetry data (filtered by user)
router.get('/', authMiddleware, async (req, res) => {
  try {
    const { user_id } = req.user;
    const { device_id, start, end, limit = 100 } = req.query;
    
    let query = `
      SELECT t.* FROM telemetry t
      WHERE t.user_id = $1
    `;
    const params = [user_id];
    let paramCount = 2;
    
    if (device_id) {
      query += ` AND t.device_id = $${paramCount}`;
      params.push(device_id);
      paramCount++;
    }
    
    if (start) {
      query += ` AND t.time >= $${paramCount}`;
      params.push(start);
      paramCount++;
    }
    
    if (end) {
      query += ` AND t.time <= $${paramCount}`;
      params.push(end);
      paramCount++;
    }
    
    query += ` ORDER BY t.time DESC LIMIT $${paramCount}`;
    params.push(parseInt(limit));
    
    const result = await db.query(query, params);
    res.json(result.rows);
  } catch (error) {
    console.error('Get telemetry error:', error);
    res.status(500).json({ error: error.message });
  }
});

// Get latest telemetry for a device
router.get('/latest/:device_id', authMiddleware, checkDeviceOwnership, async (req, res) => {
  try {
    const { device_id } = req.params;
    
    const result = await db.query(
      `SELECT * FROM telemetry 
       WHERE device_id = $1 
       ORDER BY time DESC 
       LIMIT 1`,
      [device_id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'No telemetry data found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Get latest telemetry error:', error);
    res.status(500).json({ error: error.message });
  }
});

// Get aggregated telemetry (hourly average)
router.get('/aggregate/:device_id', authMiddleware, checkDeviceOwnership, async (req, res) => {
  try {
    const { device_id } = req.params;
    const { start, end, interval = '1 hour' } = req.query;
    
    let query = `
      SELECT 
        time_bucket($1, time) AS bucket,
        AVG(temperature) AS avg_temperature,
        AVG(humidity) AS avg_humidity,
        MIN(temperature) AS min_temperature,
        MAX(temperature) AS max_temperature,
        COUNT(*) AS sample_count
      FROM telemetry
      WHERE device_id = $2
    `;
    const params = [interval, device_id];
    let paramCount = 3;
    
    if (start) {
      query += ` AND time >= $${paramCount}`;
      params.push(start);
      paramCount++;
    }
    
    if (end) {
      query += ` AND time <= $${paramCount}`;
      params.push(end);
      paramCount++;
    }
    
    query += ` GROUP BY bucket ORDER BY bucket DESC`;
    
    const result = await db.query(query, params);
    res.json(result.rows);
  } catch (error) {
    console.error('Get aggregated telemetry error:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: Server_Python/api/routes/telemetry.py
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import Optional
from services.database import db
from middleware.auth import get_current_user, check_device_ownership

router = APIRouter(prefix='/api/telemetry', tags=['telemetry'])

@router.get('/')
async def get_telemetry(
    device_id: Optional[str] = None,
    start: Optional[str] = None,
    end: Optional[str] = None,
    limit: int = Query(100, le=1000),
    current_user: dict = Depends(get_current_user)
):
    try:
        user_id = current_user['user_id']
        
        query = 'SELECT t.* FROM telemetry t WHERE t.user_id = %s'
        params = [user_id]
        
        if device_id:
            query += ' AND t.device_id = %s'
            params.append(device_id)
        
        if start:
            query += ' AND t.time >= %s'
            params.append(start)
        
        if end:
            query += ' AND t.time <= %s'
            params.append(end)
        
        query += ' ORDER BY t.time DESC LIMIT %s'
        params.append(limit)
        
        result = db.query(query, tuple(params))
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get('/latest/{device_id}')
async def get_latest_telemetry(
    device_id: str,
    current_user: dict = Depends(get_current_user),
    ownership: bool = Depends(check_device_ownership)
):
    try:
        result = db.query(
            """SELECT * FROM telemetry 
               WHERE device_id = %s 
               ORDER BY time DESC 
               LIMIT 1""",
            (device_id,)
        )
        
        if not result:
            raise HTTPException(status_code=404, detail='No telemetry data found')
        
        return result[0]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get('/aggregate/{device_id}')
async def get_aggregate_telemetry(
    device_id: str,
    interval: str = Query('1 hour'),
    start: Optional[str] = None,
    end: Optional[str] = None,
    current_user: dict = Depends(get_current_user),
    ownership: bool = Depends(check_device_ownership)
):
    try:
        query = """
            SELECT 
                time_bucket(%s, time) AS bucket,
                AVG((data->>'temperature')::float) AS avg_temperature,
                AVG((data->>'humidity')::float) AS avg_humidity,
                COUNT(*) AS sample_count
            FROM telemetry
            WHERE device_id = %s
        """
        params = [interval, device_id]
        
        if start:
            query += ' AND time >= %s'
            params.append(start)
        
        if end:
            query += ' AND time <= %s'
            params.append(end)
        
        query += ' GROUP BY bucket ORDER BY bucket DESC'
        
        result = db.query(query, tuple(params))
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))